File: python/30stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack=[]
        self.stack_min=[]

    def push(self, x: int) -> None:
        self.stack.append(x)
        if self.stack_min==[]:
            self.stack_min.append(x)
        else:
            val=min(x,self.stack_min[-1])
            self.stack_min.append(val)


    def pop(self) -> None:
        self.stack.pop() # pop(-1)
        self.stack_min.pop() # pop(-1)


    def top(self) -> int:
        return self.stack[-1]


    def min(self) -> int:
        return self.stack_min[-1]

if __name__=="__main__":
    minStack=MinStack()
    minStack.push(-2)
    minStack.push(0)
    minStack.push(-3)
    minStack.min()  # --> 返回 -3.
    minStack.pop()
    minStack.top()  #   --> 返回 0.
    minStack.min()  # --> 返回 -2.

File: cpp/30stack.cpp
#include "head.hpp"
class MinStack{
    public:
    stack<int> A,B;
    MinStack(){}
    void push(int x){
        A.push(x);
        if(B.empty() || B.top()>=x) B.push(x);
    }
    void pop(){
        if(A.top()==B.top())
            B.pop();
        A.pop();
    }
    int top(){
        return A.top();
    }
    int min(){
        return B.top();
    }

};

int main(){
    MinStack* minStack = new MinStack();
    minStack->push(-2);
    minStack->push(0);
    minStack->push(-3);
    minStack->min();   
    minStack->pop();
    minStack->top();      
    minStack->min();   
}